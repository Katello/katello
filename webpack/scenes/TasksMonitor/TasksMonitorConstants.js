export const TASKS_MONITOR_STARTED = 'TASKS_MONITOR_STARTED';
export const TASKS_MONITOR_STOPPED = 'TASKS_MONITOR_STOPPED';
export const TASKS_MONITOR_SUCCESS = 'TASKS_MONITOR_SUCCESS';
export const TASKS_MONITOR_FAILED = 'TASKS_MONITOR_FAILED';
export const TASKS_MONITOR_ALREADY_MONITORED = 'TASKS_MONITOR_ALREADY_MONITORED';
