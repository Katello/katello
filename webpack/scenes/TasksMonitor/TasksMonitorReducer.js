import Immutable from 'seamless-immutable';
import {
  TASKS_MONITOR_STARTED,
  TASKS_MONITOR_STOPPED,
  TASKS_MONITOR_SUCCESS,
  TASKS_MONITOR_FAILED,
} from './TasksMonitorConstants';

const initialState = Immutable({ });

export default (state = initialState, action) => {
  switch (action.type) {
    case TASKS_MONITOR_STARTED:
      return state.setIn([action.payload.id], {
        interval: action.payload.interval,
        intervalId: action.payload.intervalId,
        params: action.payload.params,
        tasks: [],
      });
    case TASKS_MONITOR_STOPPED:
      return state.setIn([action.payload.id, 'intervalId'], null);

    case TASKS_MONITOR_SUCCESS:
      return state
        .setIn([action.payload.id, 'tasks'], action.payload.tasks)
        .setIn([action.payload.id, 'error'], null);

    case TASKS_MONITOR_FAILED:
      return state.setIn([action.payload.id, 'error'], action.payload.error);

    default:
      return state;
  }
};
