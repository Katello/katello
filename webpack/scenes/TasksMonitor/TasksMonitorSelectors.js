export const selectTasksMonitorState = state => state.katello.tasksMonitor;

export const selectMonitor = (state, monitorId) =>
  selectTasksMonitorState(state)[monitorId];

export const selectMonitorTasks = (state, monitorId) => {
  const monitor = selectMonitor(state, monitorId);

  return monitor && monitor.tasks;
};

export const selectIsMonitorActive = (state, monitorId) => {
  const monitor = selectMonitor(state, monitorId);

  return monitor && !!monitor.intervalId;
};
