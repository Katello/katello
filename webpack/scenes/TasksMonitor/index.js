import reducer from './TasksMonitorReducer';

// eslint-disable-next-line import/prefer-default-export
export const reducers = { tasksMonitor: reducer };
