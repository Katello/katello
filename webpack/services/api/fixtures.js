export default [
  {
    path: '/katello/api/v2/products/:id/repository_sets/:repo_set_id/enable',
    searchRegex: /\/katello\/api\/v2\/products\/\d+\/repository_sets\/\d+\/enable/,
    type: 'PUT',
    response: () => {
      const types = ['yum', 'source_rpm', 'debug', 'iso', 'beta', 'kickstart'];

      const output = {
        id: Math.round(Math.random() * 500),
        name: 'Red Hat Enterprise Linux 6.2 Server for RHS 2 VSA Beta RPMs from RHUI x86_64',
        content_type: types[Math.floor(Math.random() * types.length)],
      };

      return [
        200,
        {
          id: 'd4a4432e-4987-4206-aa99-615b5b15fbf8',
          label: 'Actions::Katello::RepositorySet::EnableRepository',
          pending: false,
          username: 'admin',
          started_at: '2017-12-19 14:33:41 UTC',
          ended_at: '2017-12-19 14:33:44 UTC',
          state: 'stopped',
          result: 'success',
          progress: 1.0,
          input: {
            services_checked: ['pulp', 'pulp_auth', 'candlepin', 'candlepin_auth'],
            repository: {
              id: 7,
              name: 'Red Hat Enterprise Linux 6.2 Server for RHS 2 VSA Beta RPMs from RHUI x86_64',
              label: 'Red_Hat_Enterprise_Linux_6_2_Server_for_RHS_2_VSA_Beta_RPMs_from_RHUI_x86_64',
            },
            product: {
              id: 155,
              name: 'Red Hat Storage',
              label: 'Red_Hat_Storage',
              cp_id: '171',
            },
            provider: {
              id: 2,
              name: 'Red Hat',
            },
            organization: {
              id: 1,
              name: 'Default Organization',
              label: 'Default_Organization',
            },
          },
          output,
          humanized: {
            action: 'Enable',
            input: [],
            output: '',
            errors: [],
          },
          cli_example: null,
        },
      ];
    },
  },
  {
    path: '/katello/api/v2/products/:id/repository_sets/:repo_set_id/disable',
    searchRegex: /\/katello\/api\/v2\/products\/\d+\/repository_sets\/\d+\/disable/,
    type: 'PUT',
    response: () => [
      200,
      {
        id: 'd4a4432e-4987-4206-aa99-615b5b15fbf8',
        label: 'Actions::Katello::RepositorySet::DisableRepository',
        pending: false,
        username: 'admin',
        started_at: '2017-12-19 14:33:41 UTC',
        ended_at: '2017-12-19 14:33:44 UTC',
        state: 'stopped',
        result: 'success',
        progress: 1.0,
        input: {
          services_checked: ['pulp', 'pulp_auth', 'candlepin', 'candlepin_auth'],
          repository: {
            id: 7,
            name: 'Red Hat Enterprise Linux 6.2 Server for RHS 2 VSA Beta RPMs from RHUI x86_64',
            label: 'Red_Hat_Enterprise_Linux_6_2_Server_for_RHS_2_VSA_Beta_RPMs_from_RHUI_x86_64',
          },
          product: {
            id: 155,
            name: 'Red Hat Storage',
            label: 'Red_Hat_Storage',
            cp_id: '171',
          },
          provider: {
            id: 2,
            name: 'Red Hat',
          },
          organization: {
            id: 1,
            name: 'Default Organization',
            label: 'Default_Organization',
          },
        },
        output: {},
        humanized: {
          action: 'Disable',
          input: [],
          output: '',
          errors: [],
        },
        cli_example: null,
      },
    ],
  },
  {
    path: '/organizations/:id/repository_sets',
    searchRegex: /\/organizations\/\d+\/repository_sets$/,
    response: () => [
      200,
      {
        total: 5,
        subtotal: 5,
        page: null,
        per_page: null,
        error: null,
        search: null,
        sort: {
          by: null,
          order: null,
        },
        results: [
          {
            repositories: [],
            product: {
              name: 'Red hat Enterprise Linux Server 7',
              id: 5,
            },
            type: 'file',
            vendor: 'Red Hat',
            gpgUrl: 'http://',
            contentUrl: '/content/dist/rhel/server/7///source/iso',
            id: '2457',
            name: 'Red Hat Enterprise Linux 7 Server (Source ISOs)',
            label: 'rhel-7-server-source-isos',
          },
          {
            repositories: [],
            type: 'yum',
            product: {
              name: 'Red hat Enterprise Linux Server 7',
              id: 5,
            },
            vendor: 'Red Hat',
            gpgUrl: 'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release',
            contentUrl: '/content/dist/rhel/server/6/6Server//sat-tools/6.1/source/SRPMS',
            id: '4187',
            name: 'Red Hat Satellite Tools 6.1 (for RHEL 6 Server) (Source RPMs)',
            label: 'rhel-6-server-satellite-tools-6.1-source-rpms',
          },
          {
            repositories: [],
            type: 'yum',
            product: {
              name: 'Red hat Enterprise Linux Server 7',
              id: 5,
            },
            vendor: 'Red Hat',
            gpgUrl:
              'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-beta,file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release',
            contentUrl: '/content/beta/rhel/server/6/6Server//rh-common/os',
            id: '3076',
            name: 'Red Hat Enterprise Linux 6 Server - RH Common Beta (RPMs)',
            label: 'rhel-6-server-rh-common-beta-rpms',
          },
          {
            repositories: [],
            type: 'yum',
            product: {
              name: 'Red hat Enterprise Linux Server 7',
              id: 5,
            },
            vendor: 'Red Hat',
            gpgUrl: 'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release',
            contentUrl: '/content/dist/rhel/server/6///rhev-agent/3/os',
            id: '1699',
            name: 'Red Hat Enterprise Virtualization Agents for RHEL 6 Server (RPMs)',
            label: 'rhel-6-server-rhev-agent-rpms',
          },
          {
            repositories: [
              {
                id: 631,
                name: 'Red Hat Enterprise Linux 6 Server Kickstart x86_64 6.8',
                releasever: '6.8',
                arch: 'x86_64',
              },
            ],
            product: {
              name: 'Red hat Enterprise Linux Server 7',
              id: 5,
            },
            type: 'kickstart',
            vendor: 'Red Hat',
            gpgUrl: 'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release',
            contentUrl: '/content/dist/rhel/server/6///kickstart',
            id: '1952',
            name: 'Red Hat Enterprise Linux 6 Server (Kickstart)',
            label: 'rhel-6-server-kickstart',
          },
        ],
      },
    ],
  },
  {
    // Returns the currently enabled repositories
    path: '/repository_sets?organization_id=:id',
    searchRegex: /\/organizations\/\d+\/repository_sets/,
    response: () => [
      200,
      {
        total: 5,
        subtotal: 5,
        page: null,
        per_page: null,
        error: null,
        search: null,
        sort: {
          by: null,
          order: null,
        },
        results: [
          {
            repositories: [
              {
                id: 631,
                name: 'Red Hat Enterprise Linux 6 Server Kickstart x86_64 6.8',
                releasever: '6.8',
                arch: 'x86_64',
              },
            ],
            product: {
              name: 'Red hat Enterprise Linux Server 7',
              id: 5,
            },
            type: 'kickstart',
            vendor: 'Red Hat',
            gpgUrl: 'file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release',
            contentUrl: '/content/dist/rhel/server/6///kickstart',
            id: '1952',
            name: 'Red Hat Enterprise Linux 6 Server (Kickstart)',
            label: 'rhel-6-server-kickstart',
          },
        ],
      },
    ],
  },
  {
    path: '/products/:product_id/repository_sets/:content_id/available_repositories',
    searchRegex: /\/products\/\d+\/repository_sets\/\d+\/available_repositories/,
    response: () => [
      200,
      {
        total: 6,
        subtotal: 6,
        page: null,
        per_page: null,
        error: null,
        search: null,
        sort: {
          by: null,
          order: null,
        },
        results: [
          {
            substitutions: {
              releasever: '7.0',
              basearch: 'x86_64',
            },
            path: '/content/dist/rhel/server/7/7.0/x86_64/supplementary/os',
            repo_name: 'Red Hat Enterprise Linux 7 Server - Supplementary RPMs x86_64 7.0',
            name: 'Red Hat Enterprise Linux 7 Server - Supplementary (RPMs)',
            pulp_id:
              'Default_Organization-Red_Hat_Enterprise_Linux_Server-Red_Hat_Enterprise_Linux_7_Server_-_Supplementary_RPMs_x86_64_7_0',
            enabled: false,
            promoted: false,
          },
          {
            substitutions: {
              releasever: '7.1',
              basearch: 'x86_64',
            },
            path: '/content/dist/rhel/server/7/7.1/x86_64/supplementary/os',
            repo_name: 'Red Hat Enterprise Linux 7 Server - Supplementary RPMs x86_64 7.1',
            name: 'Red Hat Enterprise Linux 7 Server - Supplementary (RPMs)',
            pulp_id:
              'Default_Organization-Red_Hat_Enterprise_Linux_Server-Red_Hat_Enterprise_Linux_7_Server_-_Supplementary_RPMs_x86_64_7_1',
            enabled: false,
            promoted: false,
          },
          {
            substitutions: {
              releasever: '7.2',
              basearch: 'x86_64',
            },
            path: '/content/dist/rhel/server/7/7.2/x86_64/supplementary/os',
            repo_name: 'Red Hat Enterprise Linux 7 Server - Supplementary RPMs x86_64 7.2',
            name: 'Red Hat Enterprise Linux 7 Server - Supplementary (RPMs)',
            pulp_id:
              'Default_Organization-Red_Hat_Enterprise_Linux_Server-Red_Hat_Enterprise_Linux_7_Server_-_Supplementary_RPMs_x86_64_7_2',
            enabled: false,
            promoted: false,
          },
          {
            substitutions: {
              releasever: '7.3',
              basearch: 'x86_64',
            },
            path: '/content/dist/rhel/server/7/7.3/x86_64/supplementary/os',
            repo_name: 'Red Hat Enterprise Linux 7 Server - Supplementary RPMs x86_64 7.3',
            name: 'Red Hat Enterprise Linux 7 Server - Supplementary (RPMs)',
            pulp_id:
              'Default_Organization-Red_Hat_Enterprise_Linux_Server-Red_Hat_Enterprise_Linux_7_Server_-_Supplementary_RPMs_x86_64_7_3',
            enabled: false,
            promoted: false,
          },
          {
            substitutions: {
              releasever: '7.4',
              basearch: 'x86_64',
            },
            path: '/content/dist/rhel/server/7/7.4/x86_64/supplementary/os',
            repo_name: 'Red Hat Enterprise Linux 7 Server - Supplementary RPMs x86_64 7.4',
            name: 'Red Hat Enterprise Linux 7 Server - Supplementary (RPMs)',
            pulp_id:
              'Default_Organization-Red_Hat_Enterprise_Linux_Server-Red_Hat_Enterprise_Linux_7_Server_-_Supplementary_RPMs_x86_64_7_4',
            enabled: false,
            promoted: false,
          },
          {
            substitutions: {
              releasever: '7Server',
              basearch: 'x86_64',
            },
            path: '/content/dist/rhel/server/7/7Server/x86_64/supplementary/os',
            repo_name: 'Red Hat Enterprise Linux 7 Server - Supplementary RPMs x86_64 7Server',
            name: 'Red Hat Enterprise Linux 7 Server - Supplementary (RPMs)',
            pulp_id:
              'Default_Organization-Red_Hat_Enterprise_Linux_Server-Red_Hat_Enterprise_Linux_7_Server_-_Supplementary_RPMs_x86_64_7Server',
            enabled: false,
            promoted: false,
          },
        ],
      },
    ],
  },
];
