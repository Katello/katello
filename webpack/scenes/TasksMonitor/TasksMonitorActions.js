import { foremanTasksApi as api } from '../../services/api';
import { propsToSnakeCase } from '../../services/index';

import {
  TASKS_MONITOR_STARTED,
  TASKS_MONITOR_STOPPED,
  TASKS_MONITOR_SUCCESS,
  TASKS_MONITOR_FAILED,
  TASKS_MONITOR_ALREADY_MONITORED,
} from './TasksMonitorConstants';
import { selectMonitor, selectIsMonitorActive } from './TasksMonitorSelectors';

const isUnauthorizedResponse = apiResponse => apiResponse.status === 401;

const performBulkSearch = (params = {}) => {
  const apiParams = {
    searches: [{ ...propsToSnakeCase(params) }],
  };

  return api.post('/tasks/bulk_search', apiParams);
};

export const stopMonitoringTasks = id => (dispatch, getState) => {
  const { intervalId } = selectMonitor(getState(), id);

  clearInterval(intervalId);

  dispatch({
    type: TASKS_MONITOR_STOPPED,
    payload: { id, intervalId },
  });
};

export const runMonitorLifecycle = id => async (dispatch, getState) => {
  try {
    const { params } = selectMonitor(getState(), id);
    const response = await performBulkSearch(params);

    if (isUnauthorizedResponse(response)) {
      dispatch(stopMonitoringTasks(id));
      throw new Error('Unauthorized');
    }

    const tasks = response.data[0].results;

    dispatch({
      type: TASKS_MONITOR_SUCCESS,
      payload: { id, tasks },
    });
  } catch (error) {
    dispatch({
      type: TASKS_MONITOR_FAILED,
      payload: { id, error },
    });
  }
};

export const startMonitoringTasks = ({ id = 'some-uid', interval = 5000, params = {} }) => (
  dispatch,
  getState,
) => {
  if (selectIsMonitorActive(getState(), id)) {
    return dispatch({
      type: TASKS_MONITOR_ALREADY_MONITORED,
      payload: { id, interval, params },
    });
  }

  const intervalId = setInterval(() => dispatch(runMonitorLifecycle(id)), interval);

  dispatch({
    type: TASKS_MONITOR_STARTED,
    payload: {
      id, interval, intervalId, params,
    },
  });

  // run it manually so we don't wait the interval for the first time
  return dispatch(runMonitorLifecycle(id));
};
