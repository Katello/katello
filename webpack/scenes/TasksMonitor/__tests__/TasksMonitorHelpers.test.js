import { task, taskWithErrors } from './TasksMonitor.fixtures';
import { notifyTaskStartedToast, notifyTaskFinishedToast } from '../TasksMonitorHelpers';

jest.mock('../../../move_to_foreman/foreman_toast_notifications', () => ({
  notify: jest.fn(data => data),
}));

describe('TasksMonitor helpers', () => {
  it('should notify task started toast', () =>
    expect(notifyTaskStartedToast(task)).toMatchSnapshot());

  it('should notify task finished toast', () =>
    expect(notifyTaskFinishedToast(task)).toMatchSnapshot());

  it('should notify task finished toast with errors', () =>
    expect(notifyTaskFinishedToast(taskWithErrors)).toMatchSnapshot());
});
