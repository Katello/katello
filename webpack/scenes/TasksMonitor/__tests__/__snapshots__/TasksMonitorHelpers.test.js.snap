// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksMonitor helpers should notify task finished toast 1`] = `
Object {
  "message": "<span><span>Task some-action completed with a result of some-results. </span><a href=\\"/foreman_tasks/tasks/some-id/\\">Click here to go to the tasks page for the task.</a></span>",
  "type": "some-results",
}
`;

exports[`TasksMonitor helpers should notify task finished toast with errors 1`] = `
Object {
  "message": "<span><span>Task some-action completed with a result of some-results. </span><ul><li> some error number 1 </li><li> some error number 2 </li></ul><a href=\\"/foreman_tasks/tasks/some-id-with-errors/\\">Click here to go to the tasks page for the task.</a></span>",
  "type": "some-results",
}
`;

exports[`TasksMonitor helpers should notify task started toast 1`] = `
Object {
  "message": "<span><span>Task some-action has started. </span><a href=\\"/foreman_tasks/tasks/some-id/\\">Click here to go to the tasks page for the task.</a></span>",
  "type": "info",
}
`;
